include/circt/Dialect/FIRRTL/Namespace.h

```diff
@@ -59,29 +46,72 @@ class CircuitNamespace {
   /// 1. The original name is returned.
   /// 2. The name is given a `_<n>` suffix where `<n>` is a number starting from
   ///    `_0` and incrementing by one each time.
-  std::string newName(const Twine &name) {
+  StringRef newName(const Twine &name) {
     // Special case the situation where there is no name collision to avoid
     // messing with the SmallString allocation below.
     llvm::SmallString<64> tryName;
-    StringRef workingName = name.toStringRef(tryName);
-    if (internal.insert(workingName).second)
-      return workingName.str();
+    auto inserted = internal.insert(name.toStringRef(tryName));
+    if (inserted.second)
+      return inserted.first->getKey();
 
     // Try different suffixes until we get a collision-free one.
     size_t i = 0;
     if (tryName.empty())
       name.toVector(tryName); // toStringRef may leave tryName unfilled
     tryName.push_back('_');
     size_t baseLength = tryName.size();
-    do {
+    for (;;) {
       tryName.resize(baseLength);
       Twine(i++).toVector(tryName); // append integer to tryName
-    } while (!internal.insert(tryName).second);
-    return std::string(tryName);
+      auto inserted = internal.insert(tryName);
+      if (inserted.second)
+        return inserted.first->getKey();
+    }
+  }
+
+protected:
+  llvm::StringSet<> internal;
+};
+
+/// The namespace of a `CircuitOp`, generally inhabited by modules.
+class CircuitNamespace : public NamespaceBase {
+public:
+  CircuitNamespace() {}
+  CircuitNamespace(CircuitOp circuit) { add(circuit); }
+
+  /// Populate the namespace from a circuit operation. This namespace will be
+  /// composed of any operation in the first level of the circuit that contains
+  /// a symbol.
+  void add(CircuitOp circuit) {
+    for (auto &op : *circuit.getBody())
+      if (auto symbol = op.getAttrOfType<mlir::StringAttr>(
+              SymbolTable::getSymbolAttrName()))
+        internal.insert(symbol.getValue());
+  }
```

Comment:
Should this be private?

Reply:
I don't think so. There's a usage pattern in passes where you keep a `CircuitNamespace` in the pass struct, and use `clear()` and `add()` to populate it for a new circuit. This has the (incredibly small since there's only O(1) circuits) advantage of reusing the previous memory allocation for the symbols… Maybe more useful for modules?

Reply:
Sounds good to me. 👍